### ISSUE TITLE
Class generics forgotten when calling a method with .call or .apply.

### ISSUE THREAD

**Issue by dimvar** [Microsoft/TypeScript#15298]
**TypeScript Version:**

TypeScript playground

**Code**

```ts
class Foo<T> {
  myprop: T;
  method(): Foo<T> { return this; }
}

function f(x: Foo<number>) {
  var y: Foo<string>;
  y = x.method(); // correct warning here
  y = Foo.prototype.method.call(x); // missed warning here
}
```

**Expected behavior:**

When using .call or .apply on a method, the method type is turned to a function type whose first argument is the receiver object of the method. If the class of the receiver is generic, the type variables should be added to the function type.

**Actual behavior:**

The class generics become any (I assume), and the type error is not detected.

**Comment by jesseschalken** [comment]
`call` and `apply` are not type safe generally.

```ts
function foo(x: string): number {
    return 1;
}

foo.call(4, 8, 2, 4, 5); // ok
foo.apply(4, [2]); // ok
```

**Comment by DanielRosenwasser** [comment]
Hey @dimvar, this looks like a duplicate of #212. We don't special case the behavior of any of those methods. Does Closure have a generalized declaration that could work for functions other than `call`, `bind`, and `apply`?

**Comment by mhegazy** [comment]
This is not rely a duplicate of #212; though you need #212 for this to behave as the OP suggests.

The issue here is the `prototype` of a generic class is an instantiation with type `any`.  so `Foo.prototype` is `Foo<any>`.

**Comment by dimvar** [comment]
> The issue here is the prototype of a generic class is an instantiation with type any. so Foo.prototype is Foo<any>.

Exactly. I was fixing this in the Closure type checker and ran into a similar issue.

**Comment by mhegazy** [comment]
This issue seems to have been already addressed or is unactionable at the moment. I am auto-closing it for now.

**Comment by dimvar** [comment]
Are you referring to the command-line TypeScript compiler? I just checked and the issue still happens when using the playground.

**Comment by mhegazy** [comment]
this issue is not fixed, it is a design limitation that we have always had in the system. the message above is my auto-close script.
